lib/graphql/metrics/instrumentation.rb

```diff
@@ -0,0 +1,48 @@
+# frozen_string_literal: true
+
+module GraphQL
+  module Metrics
+    class Instrumentation
+      def before_query(query)
+        unless query_present_and_valid?(query)
+          query.context[GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS] = true
+        end
+
+        # NOTE: This context value may have been set to true in the application, so we should still return early here if
+        # it's set, even if the query is valid.
+        return if query.context[GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS]
+
+        ns = query.context.namespace(CONTEXT_NAMESPACE)
+        ns[GraphQL::Metrics::TIMINGS_CAPTURE_ENABLED] = true
+        ns[GraphQL::Metrics::INLINE_FIELD_TIMINGS] = {}
+        ns[GraphQL::Metrics::LAZY_FIELD_TIMINGS] = {}
+      end
+
+      def after_query(query)
+        return if query.context[GraphQL::Metrics::SKIP_GRAPHQL_METRICS_ANALYSIS]
+
+        ns = query.context.namespace(CONTEXT_NAMESPACE)
+        query_duration = GraphQL::Metrics.current_time_monotonic - ns[GraphQL::Metrics::QUERY_START_TIME_MONOTONIC]
```

Comment:
worth a comment here I think that we start the clock even before the instrumentation's `before_query`, this keeps catching me